### ISSUE TITLE
Confusing aspects of Group admin guide

### ISSUE THREAD

**Issue by bhaugen** [open-app/holodex#70]
These might want to become separate issues, but I am plopping them together for now...
- As far as I can tell, https://github.com/holodex/enspiral-data does not have a roleTypes directory. Is one necessary? 
  - See also https://github.com/open-app/holodex/issues/63
- Is there a principle behind what is source and what is target in relationshipTypes? Examples:
  - set: source: subgroup, target: parent-group
  - membership: source: member, target: group
  - stewardship: source: steward, target: stewardee
  - Note: we have struggled with this issue, too. RelationshipTypes need to be very generic, so the names of the "ends" need to be generic as well, and generic names don't give much direction for implementors.
- What's the estimated timing on https://github.com/open-app/holodex/issues/69 ? 
  - Might affect whether I cheat and write YAML by hand or generate it programmatically.
- Do you guys intend to host a demo that could display other org data, or should we plan on running the holodex app ourselves (this is in the short term)?

**Comment by ahdinosaur** [comment]
- we don't have any standalone role types yet, so we put our role types within the relationship types
- there's no reason both sides can't be both source and target, which i reckon is true of parent / child and group / member relationships
- good question, i've warming up more and more to doing it, so this add more fuel to that fire. what's the tradeoff between writing by hand and generating programmatically? we could be keen to help y'all generate programmatically given a raw dump
- yes, we can host an instance for y'all

**Comment by bhaugen** [comment]
Thanks. Especially about the instance. Can you host an instance that different networks could plug into by means of a query parameter or something? We got more than one network, and surely others will follow in our footsteps as soon as they see the delicious eye candy.

We want to generate programmatically for production use. And have holodex consume the output of an API rather than github files, but we can wait on that.

Re source and target: I don't see the relationships duplicated from both the source and target sides. Looks like they are only recorded from the source side. Did I miss something?

(We don't care, by the way. We record the relationships in their own table, with links to the agents on both ends. But we are happy to conform to what holodex wants, in order to get that snazzy UI.)

**Comment by ahdinosaur** [comment]
> Can you host an instance that different networks could plug into by means of a query parameter or something? We got more than one network, and surely others will follow in our footsteps as soon as they see the delicious eye candy.

it's possible we fit everyone onto a single instance (allowing for multiple sources of data), or we make it easy to host multiple instances. we'll see what makes the most sense when it happens.

> Re source and target: I don't see the relationships duplicated from both the source and target sides. Looks like they are only recorded from the source side. Did I miss something?

the YAML might be confusing, but basically the data is a flat list of typed objects. in the YAML, any nesting is flattened (and [maybe a foreign key added](https://github.com/holodex/app/blob/master/src/node_modules/api/index.js#L87-L94)) when it's read.

**Comment by bhaugen** [comment]
That didn't answer my question about how Holodex wants relationships to be recorded in YAML.

Here's an excerpt of mj-kaplan.yml:

```
'@type': Person
name: MJ
me: &me
  agent: people/mj-kaplan

relationships:
  - '@type': Relationship
    type: relationshipTypes/stewardship
    context: groups/loomio
    roles:
      - '@type': Role
        type: roleTypes/steward
        <<: *me 
      - '@type': Role
        type: roleTypes/stewardee
        agent: people/matthew-bartlett
```

That same relationship does not appear in matthew-bartlet.yml (the stewardee). Instead, I see this other relationship where Matthew is the steward.

```
'@type': Person
name: Matthew Bartlett
me: &me
  agent: people/matthew-bartlett

relationships:
  - '@type': Relationship
    type: relationshipTypes/stewardship
    context: groups/loomio
    roles:
      - '@type': Role
        type: roleTypes/steward
        <<: *me 
      - '@type': Role
        type: roleTypes/stewardee
        agent: people/jon
```

In other words, it looks to me like you are recording each relationship once, in the yml file for the source. Correct? Or am I still missing something?

**Comment by simontegg** [comment]
That's correct. It only needs to be recorded once (at this stage).

**Comment by bhaugen** [comment]
Thanks, got it. Can do.

**Comment by bhaugen** [comment]
LinkTypes: some relationshipTypes have one, some have two. What's the difference?

set.yml:

```
links:
  - '@type': LinkType
    source: subgroup
    target: parent-group
    label: "are part of {target}"
```

membership.yml:

```
links:
  - '@type': LinkType
    source: member
    target: group
    label: "belong to {target}"
  - '@type': LinkType
    source: group
    target: member
    label: "include {target}"
```

**Comment by simontegg** [comment]
Sorry Bob. We're still figuring this out. The one above is actually an error. 

Although it seems from the user testing Mikey's been doing that some relationships have a primary direction. for stewardship this is `steward ->[stewards]-> stewardee`. Users found the `stewardee ->[are-stewarded-by]-> steward` confusing and superfluous. In the interests of simplifying the UI we're making links in both directions optional. 

I've just pushed my development branch which has my latest thinking https://github.com/holodex/enspiral-data/tree/links/relationshipTypes

Since this hasn't settled, don't worry too much about meeting a particular format. I'm happy to massage it into the right format before deploy.

**Comment by bhaugen** [comment]
Thanks. I'm not very worried, but am going back and forth between our data, which when I think about publishing in the holodex format, needs to be cleaned up some, and rethinking our relationshipTypes in terms of holodex roleTypes, which we don't have. But which I will add for this export.

Lotsa fun. Going slow, thinking. Should have something out in json in a couple of days, if I don't get bogged down in Sensorica work too much.

**Comment by bhaugen** [comment]
I'm still muddled in mapping between our model and yours. They are very similar, but also different, sometimes subtly, and sometimes not so subtly.

I will write more about this when I think I understand better. But in the meantime, do you guys have anything beyond the yaml files and the admin guide, like a model or schema or picture? If not, no worries, don't create one for me, I'll draw my own pictures.

**Comment by ahdinosaur** [comment]
- [diagram and psuedo-schema that's close to the current setup](https://github.com/openvocab/holodex/issues/1#issuecomment-81310624)
- [slightly more up-to-date (but possibly confusing) schemas](https://github.com/openvocab/holodex/tree/master/vocab)

i'm keen to produce an updated diagram and psuedo-schema with our current vocab, as we are solidifying our models after adding `Link`, `LinkType`, and `Query` and [i want to start using the vocab proper in our app](https://github.com/open-app/holodex/issues/72).

**Comment by ahdinosaur** [comment]
@bhaugen here's a quick UML diagram i made, hope it helps:

![holodex vocab UML](http://i.imgur.com/bphOWeB.jpg)

`Query`, `Link`, and `LinkType` types are still in progress.

**Comment by bhaugen** [comment]
Thanks. Very helpful.

**Comment by bhaugen** [comment]
Back on the job. Here's a mapping from our vocab to yours. Note gaps.

![holodex_uml](https://cloud.githubusercontent.com/assets/117439/8662863/f1a27d7c-2989-11e5-9fe8-89706cf966c1.jpg)

Main gaps seem to be Roles and Role Types. We'll fake it for now. 

Question: we got AgentTypes, which you don't have. You got Person and Group, which to us are AgentTypes. So you think RoleTypes would be relates to AgentTypes? As in, this Role can be performed  by a Person, and that Role can be performed by a Group?

More to come, as in json data.

**Comment by ahdinosaur** [comment]
@bhaugen 

sweet! nice work. :)

we don't have AgentTypes at the moment, but maybe we should. right now Person and Group are classically inherited from Agent.

**Comment by bhaugen** [comment]
I'll start implementing today. For github repos, do you care if I do one repo for all OVN networks, with subdirectories per net, or do you want a separate repo per net?

**Comment by fosterlynn** [comment]
@ahdinosaur  I'm joining @bhaugen in getting deeper into your model.  Fun!  :)  

I have a question: Under what conditions would there be more than one LinkType for a RoleType, or more than one Link for a Role?  And examples?

**Comment by fosterlynn** [comment]
@ahdinosaur Follow-up, having studied the data a bit more: It looks more like a LinkType has 2 RoleTypes always.  You might have one LinkType in a RelationshipType if it only makes sense to traverse the relationship in one direction in the app; if it makes sense in both directions, you have two LinkTypes.  Is this getting closer?

(My earlier question was based on the picture.)

**Comment by fosterlynn** [comment]
@ahdinosaur Still looking at it..... realizing you are asleep, sorry about the continuing thought process, hope it isn't too spammy....  

OK, a few more questions: 
- Is it true that a RoleType can't be reused in another RelationshipType?  I assume not.
- Do you really need Link?  Or can you just use the LinkType and RoleType info?
- Should context agent be in Relationship?  Actually, I see this in your data, makes sense to me.
- Can a Role really not have a Relationship?  If so, examples?

So, I took the liberty of drawing a new picture, which has some assumptions in answers to my questions, so might not be right..... but maybe it is helpful towards getting us outputting data to you, I hope....
![holodex uml my version](https://cloud.githubusercontent.com/assets/3776081/8684673/5ee3ec6c-2a3f-11e5-946f-9656c45a5022.png)

**Comment by ahdinosaur** [comment]
> For github repos, do you care if I do one repo for all OVN networks, with subdirectories per net, or do you want a separate repo per net?

one repo for all networks is good.

> You might have one LinkType in a RelationshipType if it only makes sense to traverse the relationship in one direction in the app; if it makes sense in both directions, you have two LinkTypes. Is this getting closer?

yes, you got it.

> Is it true that a RoleType can't be reused in another RelationshipType? I assume not.

i'm not really sure about this, we haven't reused a RoleType yet but theoretically you could?

> Do you really need Link? Or can you just use the LinkType and RoleType info?

we haven't actually started using Links yet, they are something on our horizon that we're exploring how they fit in.

> Should context agent be in Relationship? Actually, I see this in your data, makes sense to me.

yes, sweet.

> Can a Role really not have a Relationship? If so, examples?

i think so. an example off the top of my head would be a Treasurer of a group. it's a role with a context agent, but not quite a relationship. this is another thing we haven't fully explored yet. 

> So, I took the liberty of drawing a new picture...

woo! looks great. :smile: 

feel free to continue asking questions, i'm happy to help as best i can, y'all are really awesome for collaborating with us. :revolving_hearts:

**Comment by fosterlynn** [comment]
@ahdinosaur Hi y'all, we're again back in the saddle.  Bob and I are throwing up the holodex work like a hot potato and now I have it.  (Just kidding of course!  We're fighting over it and I won this morning!) :smile: 

First question: in the agent file, the agent relationships, the roles for the agent relationship, is agent some kind of key to the agent, or is it a location, or something else?  (For example: agent: people/mikey, agent: groups/enspiral-craftworks.)  Should we use a URL to our agent page?  Or should we end up with a directory structure of data like yours and reference that?

EDIT: Decided to use your directory structure for this.  Seems like everything needs to be put into that.....

**Comment by fosterlynn** [comment]
Another question (and input for the admin guide): How should users think about creating the RoleType and LinkType labels?

**Comment by ahdinosaur** [comment]
> First question: in the agent file, the agent relationships, the roles for the agent relationship, is agent some kind of key to the agent, or is it a location, or something else? (For example: agent: people/mikey, agent: groups/enspiral-craftworks.) Should we use a URL to our agent page? Or should we end up with a directory structure of data like yours and reference that?

yes, `agent` is an 'id' of an agent, where the 'id' of an object is by default the directory path to that object of the form `path/to/file#object/properties`.

> Another question (and input for the admin guide): How should users think about creating the RoleType and LinkType labels?

cc @simontegg

**Comment by simontegg** [comment]
@fosterlynn 

In my latest pull request I've used the concept of `labels` and  `queryLabels` https://github.com/holodex/enspiral-data/pull/21/files

These only really apply to LinkTypes currently, but will also apply to RoleTypes.

Plain old `labels` are descriptions of the link, as phrases and _as you would expect to see it as an annotation on the graph_ https://twitter.com/EdwardTufte/status/622786266690646016 . For example, [Charlie] -> [stewards] -> [Simon] or [Simon] -> [belongs to] -> [Craftwoks]. The general form (in English) is `{source} [verb] {target}`, but noun-y role-y descriptions can be ok if that's what makes sense.

To avoid enforcing our preferences on users a `linkType` `has-many` `labels` and  `has-many` `queryLabels` , and can be referenced as a string with comma delimiter (short-term hack).

`queryLabels` are used when we use the linkType as the basis for a Query (which doesn't officially exist yet). A Query is the combination of a linkType plus a `source` | `target` | `context`. A queryLabel describes the 'missing' agents. For examples:

`/linkTypes/stewards``+``source:/people/simon``. ,  in plain english is "People stewarded by Simon". 

`/linkTypes/stewards` + `context:/groups/craftworks` -> "People who steward in Craftworks",  "Stewards in Craftworks" is probably more natural for this case but this will be `/roleTypes/steward` + `context:/groups/craftworks`

`/linkTypes/belongs-to``+``target:/groups/craftworks`` -> "People who belong to Craftworks" or "People in Craftworks"

For the models the way to write these is to think of some natural English sentences that describe people with a relationship to or in a `source` | `target` | `context`

 -> "People in {target},People who belong to {target}"

-> "People who steward {target}, People whom {source} stewards, People stewarded by {source}, People who steward in {context}"

Note we're going to bring `membership` back together for both people and groups after we split it into `membership` (people) and `set` (groups). This means that some queryLabels will also have an `agentType` i.e "{agentType(s)} that belong to {target},{agentType(s)} that are part of {target}", 

Note the inverse queryLabel here "Groups that include {target},Groups that incorporate {target}"

If there is no natural sounding sentence for a particular direction the linkType itself can probably be dropped.
